source/celview.cpp

```diff
@@ -214,17 +214,17 @@ void CelView::removeFrames(int startingIndex, int endingIndex)
 
 void CelView::sendAddFrameCmd(IMAGE_FILE_MODE mode, int index, const QString &imagefilePath)
 {
-    AddFrameCommand *command;
+    std::unique_ptr<Command> command;
```

Comment:
This pattern had me a pretty confused. You pass the naked pointer into `QObject::connect()` but don't relinquish ownership so it looks like the pointer gets deleted when the `command` variable goes out of scope. It seems you're passing the unique pointer into `undoStack->push()` by reference and then using `std::move()` to relinquish ownership of the reference in that function, which is very misleading to the point where I'm surprised the language even lets you do that.

It would be better to use `std::unique_ptr<AddFrameCommand>` here. Then we don't need to `dynamic_cast` the pointer in the call to `QObject::connect()`, and we can explicitly relinquish ownership using `undoStack->push(std::move(command))`. Now `UndoStack::push(std::unique_ptr<Command> command)` can take the parameter by value, and it's clear that ownership has been transferred to it.

Reply:
Thanks, changed to what you've suggested. Now it looks a lot clearer :)